html/modules/custom/ghi_plans/src/Plugin/ParagraphHandler/PlanEntityTypes.php

```diff
@@ -0,0 +1,469 @@
+        '#suffix' => '</div><br />',
+      ];
+
+      $subform['entity_ids'] = [
+        '#type' => 'tableselect',
+        '#header' => [
+          'id' => $this->t('ID'),
+          'description' => $this->t('Description'),
+        ],
+        '#options' => $entity_options,
+        '#default_value' => !empty($defaults['entity_ids']) ? array_combine($defaults['entity_ids'], $defaults['entity_ids']) : [],
+        '#prefix' => '<div id="' . $wrapper_id . '">',
+        '#suffix' => '</div>',
+        '#empty' => $this->t('No suitable plan entities found. If you save this form like this, the block will not be displayed.'),
+      ];
+
+      if (count($matching_entities)) {
+        $validation_options = $defaults;
+        $validation_options['entity_ids'] = NULL;
+        foreach ($matching_entities as $entity) {
+          if ($this->validatePlanEntity($entity, $validation_options)) {
+            continue;
+          }
+          $subform['entity_ids'][$entity->id]['#disabled'] = TRUE;
+        }
+      }
+    }
+  }
+
+  /**
+   * Ajax callback to update the entities list.
+   *
+   * @param array $form
+   *   Form array.
+   * @param \Drupal\Core\Form\FormStateInterface $form_state
+   *   Form state interface.
+   *
+   * @return mixed
+   *   The part of the form array that should be updated.
+   */
+  public function updateEntitiesList(array &$form, FormStateInterface $form_state) {
+    $triggering_element = $form_state->getTriggeringElement();
+    $parents = $triggering_element['#array_parents'];
+    array_pop($parents);
+    $parents[] = 'entity_ids';
+    return NestedArray::getValue($form, $parents);
+  }
+
+  /**
+   * Get options for the entity type dropdown.
+   *
+   * @return array
+   *   An array with valid options for the current context.
+   */
+  private function getEntityTypeOptions() {
+    $plan_id = $this->parentEntity->field_original_id->value;
+
+    if (!$plan_id) {
+      // Without a plan context, we show a default set of availabe plan entity
+      // types.
+      $default_options = [
+        'CA' => $this->t('Cluster Activities'),
+        'CO' => $this->t('Cluster Objectives'),
+        'CQ' => $this->t('Humanitarian Consequences'),
+        'SO' => $this->t('Strategic Objectives'),
+        'SP' => $this->t('Specific Objectives'),
+        'SSO' => $this->t('Sub Strategic Objectives'),
+        'OC' => $this->t('Outcomes'),
+        'OP' => $this->t('Outputs'),
+      ];
+      return $default_options;
+    }
+
+    $matching_entities = $this->getPlanEntities();
+
+    $options = [];
+    if (empty($matching_entities)) {
+      return $options;
+    }
+    $weight = [];
+    foreach ($matching_entities as $entity) {
+      $ref_code = $entity->entityPrototype->refCode;
+      if (empty($options[$ref_code])) {
+        $name = $entity->entityPrototype->value->name->en->plural;
+        $options[$ref_code] = $name;
+        $weight[$ref_code] = $entity->entityPrototype->orderNumber;
+      }
+    }
```

Comment:
It might be cleaner to return only the data we need for this, similar like what's done for `getPlanWebContentAttachments`

Reply:
Yes, ok.